### PR TITLE
Follow up on PR #17894 on 7.1 version parsing

### PR DESCRIPTION
**Pull request by OmarTawfik**
Responds to feedback on #17894 

cc @jcouv @dotnet/roslyn-compiler @dotnet/roslyn-compat-council

### CODE REVIEW

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 1]
```diff
@@ -1347,6 +1320,28 @@ public void LanguageVersion_DisplayString()
         }
 
         [Fact]
+        public void LanguageVersion_GetErrorCode()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: Thanks :-)

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 2]
```diff
@@ -1188,142 +1188,115 @@ public void ArgumentParsing()
...
+        [InlineData("3.0", LanguageVersion.CSharp3)]
+        [InlineData("4", LanguageVersion.CSharp4)]
+        [InlineData("4.0", LanguageVersion.CSharp4)]
+        [InlineData("5", LanguageVersion.CSharp5)]
+        [InlineData("5.0", LanguageVersion.CSharp5)]
+        [InlineData("6", LanguageVersion.CSharp6)]
+        [InlineData("6.0", LanguageVersion.CSharp6)]
+        [InlineData("7", LanguageVersion.CSharp7)]
+        [InlineData("7.0", LanguageVersion.CSharp7)]
+        [InlineData("7.1", LanguageVersion.CSharp7_1)]
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: Nice. Do you plan do to similar refactoring on VB side?

Comment (OmarTawfik) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: I'd rather keep this PR focused on the C# version side. Every now and then I find similar mega-tests and I break them down, so I prefer doing them as we go.
If you think there is a wider-spread and they should be dealt with sooner, I can file an issue and start looking into them.

Comment (jcouv) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: Sounds good.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 3]
```diff
@@ -4330,25 +4330,25 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
...
-    <value>Feature '{0}' is not available in C# 5.  Please use language version {1} or greater.</value>
+    <value>Feature '{0}' is not available in C# 5. Please use language version {1} or greater.</value>
   </data>
   <data name="ERR_FeatureNotAvailableInVersion6" xml:space="preserve">
-    <value>Feature '{0}' is not available in C# 6.  Please use language version {1} or greater.</value>
+    <value>Feature '{0}' is not available in C# 6. Please use language version {1} or greater.</value>
   </data>
   <data name="ERR_FeatureNotAvailableInVersion7" xml:space="preserve">
-    <value>Feature '{0}' is not available in C# 7.  Please use language version {1} or greater.</value>
+    <value>Feature '{0}' is not available in C# 7. Please use language version {1} or greater.</value>
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Why did we remove all of the spaces here?  Seems unrelated to this PR.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 4]
```diff
@@ -5033,7 +5033,10 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
     <value>Invalid name for a preprocessing symbol; '{0}' is not a valid identifier</value>
   </data>
   <data name="ERR_FeatureNotAvailableInVersion7_1" xml:space="preserve">
-    <value>Feature '{0}' is not available in C# 7.1.  Please use language version {1} or greater.</value>
+    <value>Feature '{0}' is not available in C# 7.1. Please use language version {1} or greater.</value>
+  </data>
+  <data name="ERR_LanguageVersionCannotHaveLeadingZeroes" xml:space="preserve">
+    <value>Specified language version '{0}' cannot have leading zeroes</value>
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Did we discuss what happens with trailing 0s?  Sorry if I missed that part of the email.

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 5]
```diff
@@ -805,6 +805,12 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
                             {
                                 AddDiagnostic(diagnostics, ErrorCode.ERR_SwitchNeedsString, MessageID.IDS_Text.Localize(), "/langversion:");
                             }
+                            else if (value.StartsWith("0"))
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Please use an explicit comparer here.  The default for `StartsWith` is to use culture sensitive comparisons and I'm unsure if that is correct here.  Believe this should be ordinal.

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 3 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: This is a follow-up on the last PR. `ERR_FeatureNotAvailableInVersion7_1` was added and it had this extra space. Then Omar pointed out all the previous versions had the same typo.

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 4 above)
Comment (OmarTawfik) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Per last email from @gafter, we are accepting this as a breaking change, and putting it in the Post VS2017 breaking change markdown (edited in this PR).

---
In reply to: [107026072](https://github.com/dotnet/roslyn/pull/17990#discussion_r107026072) [](ancestors = 107026072)

Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Yes but that covers only a single trailing 0.  What happens when I enter "7.000"?

The code gives an error for leading zero but not a trailing zeros.  Possibly a non-issue but want to make sure it was explicitly considered.

Comment (OmarTawfik) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: We never accepted them in the first place, so this shouldn't be a breaking change. Am I missing something?

---
In reply to: [107036325](https://github.com/dotnet/roslyn/pull/17990#discussion_r107036325) [](ancestors = 107036325)

Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: This isn't about a breaking change, it's about symmetry.  

``` txt
/langversion:07
Error: leading zeros not allowed 
/langversion:7.00
Invalid language version 
```

Seems like this could be a reasonable typo in the new point release world.  

Not hard feedback.  Just an item I want to make sure we've considered.

Comment (OmarTawfik) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: We never accepted trailing zeroes, so that's not a breaking change. Am I missing something?

Comment (OmarTawfik) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Mmm. good point.

---
In reply to: [107037475](https://github.com/dotnet/roslyn/pull/17990#discussion_r107037475) [](ancestors = 107037475)
